src/savecompiler/dypgen/doc/dypgen.1
.TH DYPGEN 1

.SH NAME
dypgen \- The Objective Caml parser generator

.SH SYNOPSIS
.B dypgen 
[
.I options
]
.I filename.dyp

.SH SYNOPSIS
.B dypgen.opt
[
.I options
]
.I filename.dyp

.SH SYNOPSIS
.B dyp2gram
.I filename.dyp

.SH DESCRIPTION

The
.BR dypgen (1)
command produces a parser from a context-free grammar
specification with attached semantic actions, in the style of
.BR yacc (1).
Assuming the input file is
.IR grammar \&.dyp,
running
.B dypgen
produces Caml code for a parser in the file
.IR grammar \&.ml,
and its interface in file
.IR grammar \&.mli.

The generated module defines one parsing function per entry point in
the grammar. These functions have the same names as the entry points.
The lexer (lexical analyzer) can be generated by dypgen or by a third
party generator such as
.BR ocamllex (1)
.
If an external lexer generator is used, then parsing functions take
as arguments a lexical analyzer (a function
from lexer buffers to tokens) and a lexer buffer. If dypgen is used
to generate the lexer then parsing functions only take a lexer buffer as argument.
They return the semantic attribute of the corresponding entry point.
Lexer buffers are an abstract data type
implemented in the library module Dyp. Tokens are values from
the concrete type token, defined in the interface file
.IR grammar \&.mli
produced by
.BR dypgen (1).

.BR dypgen.opt (1)
is the same as .BR dypgen (1) but compiled to native code instead of bytecode.

.BR dyp2gram (1)
is is a pearl script producing a readable text documentation of your grammar from a 
.IR grammar \&.dyp
file.

.SH OPTIONS

The
.BR dypgen (1)
command recognizes the following options:
.TP
.B --merge-warning prefix
 activates merge warning: the generated parser will emit a warning on the standard output each time a merge happens
.TP
.B  --pv-obj
the type constructor obj is made as a sum of polymorphic variants instead of a sum of constructors. This is useful when the maximum number of constructors allowed is reached.
.TP
.B  --pv-token
the type token is made as a sum of polymorphic variants instead of a sum of constructors. This is useful when the maximum number of constructors allowed is reached.
.TP
.B  --noemit-token-type
the type token is not emitted in the mli or ml files, it must be provided by the user instead.
.TP
.B  --no-pp
the value pp is not stated in the mli file.
.TP
.B  --no-obj-type
the type obj_type is not stated in the mli file.
.TP
.B  --no-undef-nt
prevents the exception Undefined_nt from being raised.
.TP
.B  --ocamlc "options"
dypgen uses these options when calling ocamlc -i.
.TP
.B  --command "command"
dypgen uses this command instead of calling ocamlc -i.
.TP
.B  --no-mli
dypgen does not generate the .mli file.
.TP
.B  --use-rule-order
dypgen uses rule order to decide which rule to reduce in case of reduce/reduce conflict, it is overriden by the variable dypgen_use_rule_order.
.TP
.B  --use-all-actions
the parser will execute all the user actions bound to a rule instead of just the first that does not raise Giveup.
.TP
.B  --cpp dypgen
calls the C preprocessor cpp on the input file before processing it.
.TP
.B  --cpp-options "options"
dypgen calls the C preprocessor cpp with options "options" on the input file before processing it.
.TP
.B  --version
gives the version of dypgen and exit.
.TP
.B  -help 
Display this list of options
.TP
.B  --help
Display this list of options


.SH SEE ALSO
.BR ocamllex (1).
.br
.IR "dypgen User's Manual"


fbuild/NEWS.markdown
I'm pleased to announce FBuild version 0.2. It's been way too long since the
last FBuild release, but a lot of features have been added over the past year.
Overall, the core concept of FBuild is still the same, but there have been
some structural changes in response to some suggestions I've had.

The most obvious change is that there is now a context object that you must
pass around in order to interact with cacheable functions. Now, to write a
simple builder, you must do:

    import fbuild.builders.c

    def build(ctx):
        shared = fbuild.builders.c.guess_shared(ctx)
        shared.build_exe('foo', ['foo.c', 'bar.c'])

Next is the support for command line targets. This allows you to easily call a
specifically marked function, as in:

    import fbuild.builders.ocaml
    def build(ctx):
        ocamlc = fbuild.builders.ocaml.Ocamlc(ctx)
        ocamlc.build_exe('foo', ['foo.ml'])

    import fbuild.target
    @fbuild.target.register(name='run-test', help='run the foo test')
    def run_test(ctx):
        # make sure 'foo' is built
        build(ctx)
        ctx.execute([ctx.buildroot / 'foo', 'world'])
        ctx.execute([ctx.buildroot / 'foo', 'fbuild'])

Which can be run by:

    % fbuild run-test
    looking for program ocamlc.opt : ok /opt/local/bin/ocamlc.opt
    looking for program ocamldep.opt : ok /opt/local/bin/ocamldep.opt
    determining platform             : {'bsd', 'darwin', 'macosx', 'posix'}
    checking if ocamlc.opt can make objects : ok
    checking if ocamlc.opt can make libraries: ok
    checking if ocamlc.opt can make exes    : ok
    checking if ocamlc.opt can link lib to exe: ok
     * ocamldep.opt                         : foo.ml
     * ocamlc.opt                           : build/foo.ml -> build/foo.cmo
     * ocamlc.opt                           : build/foo.cmo -> build/foo
    Hello world
    Hello fbuild!

One nice feature of the targets is that they are automatically integrated into
the "--help" option:

    % fbuild -h
    Usage: fbuild-light [options] target1 [target2 target3 ...]

    ...

    Targets:
      build
      run-test	run the foo test

Next up is another removal of magic, this time in `fbuild.config.c`. Now, if
you want to define a config test that requires a header, you must explicitly
define it yourself, as in:

    import fbuild.config.c as c
    class assert_h(c.Test):
        header = c.header_test('assert.h')

Finally, the last major addition was support for Linux's LD\_LIBRARY\_PATH and
OS X's DYLD\_LIBRARY\_PATH for shared libraries. Now it's much easier to
execute test code that depends on a shared library. To use it, just do:

    ctx.execute(cmd, runtime_libpaths=['lib1', 'lib2'])

Beyond that, we added builders for ghc, avr-gcc, gcc/iOS, O'Caml Findlib,
and O'Caml Batteries. For configuration support, we now can check for SDL,
LLVM, GNU Readline, GNU GMP, Google Test Framework, and OpenSSL. And, of
course, lots of miscellaneous bug fixes. For a more detailed breakdown, please
look at our git history.

Oh and one last thing. FBuild-0.2's build/ directory is incompatible with
FBuild-0.1's build/ directory, so you'll have to remove the old directory
before you can compile.

Please let us know if you run into any problems or need any help. You can:

 * File a bug at http://github.com/felix-lang/fbuild/issues
 * Contact our mailing list at http://groups.google.com/group/fbuild
 * Or find us on IRC on freenode.net in #felix


buildsystem/dypgen.py
import fbuild
import fbuild.db
from fbuild.path import Path

# ------------------------------------------------------------------------------

class Builder(fbuild.db.PersistentObject):
    def __init__(self, ctx, exe):
        super().__init__(ctx)

        self.exe = exe

    @fbuild.db.cachemethod
    def __call__(self, src:fbuild.db.SRC, *,
            buildroot=None,
            flags=[]) -> fbuild.db.DST:
        buildroot = buildroot or self.ctx.buildroot

        # first, copy the src file into the buildroot
        src_buildroot = src.addroot(buildroot)
        dst = src_buildroot.replaceext('.ml')

        if src != src_buildroot:
            src_buildroot.parent.makedirs()
            src.copy(src_buildroot)
            src = src_buildroot

        cmd = [self.exe, '--no-mli']
        cmd.extend(flags)
        cmd.append(src)

        self.ctx.execute(cmd, self.exe.name, '%s -> %s' % (src, dst),
            color='yellow')

        return dst

# ------------------------------------------------------------------------------

def build_lib(phase):
    path = Path('src/compiler/dypgen/dyplib')
    return phase.ocaml.build_lib(path/'dyp', Path.glob(path/'*.ml{,i}'))

def build_exe(phase):
    path = Path('src/compiler/dypgen/dypgen')
    exe = phase.ocaml.build_exe(path/'dypgen', Path.globall(
        path/'*.ml{,i}',
        phase.ocamllex(path/'dypgen_lexer.mll'),
        phase.ocamllex(path/'extract_type.mll'),
        phase.ocamllex(path/'insert_linenum.mll')),
        libs=[build_lib(phase)])

    return Builder(phase.ctx, exe)
